### PR TITLE
mtmd: support DeepSeek-V4-Flash-Vision-Exp

### PR DESCRIPTION
**Pull request by ngxson**
## Overview

> [!IMPORTANT]
>
> Text model need to be reconverted, need to apply this: https://github.com/ggml-org/llama.cpp/pull/28154

Add support for https://huggingface.co/deepseek-ai/DeepSeek-V4-Flash-Vision-Exp (text model is already supported out of the box)

Most of the things are trivial, with only one small catch: because the text model "compress" a block of 4 tokens, the image need to be aligned to this boundary (left-padded with a learned embedding token)

We assume the whole prompt is always fed into `mtmd_tokenize` for now, so that both the text and image can be counted and thus, the padding can also be counted correctly. For example: if we have 5 text tokens followed by an image, then 3 pad tokens must be added (so that image started at 5+3 = 8, aligned to 4)

## Testing

I only tested with a small slice of the real weight

| config                        | NMSE (IMAGE rows) | sentinels/layout |
|-------------------------------|-------------------|------------------|
| bf16 mmproj, Metal (default)  | 1.19e-04          | bit-exact        |
| f32 mmproj, Metal             | 1.42e-04          | bit-exact        |
| f32 mmproj, CPU               | 1.34e-05          | bit-exact        |
| f32 mmproj, CPU, -fa off      | 1.20e-05          | bit-exact        |
| truncated to 1 / 4 ViT layers | 5.7e-07           | bit-exact        |

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: ~90% of the code is AI-generated

### CODE REVIEW

**Comment by am17an** [comment]
I think this would break when we have image + text in the same ubatch because of how we handle the the kv-cache i.e. positions are conflated with cells

**Comment by ngxson** [comment]
@am17an not sure what you mean, but currently we cannot have both image embd + text in batch (or ubatch). that would require https://github.com/ggml-org/llama.cpp/pull/24669 first

the only case where embd + text in same `llama_batch` that we currently support is MTP embd

**Comment by am17an** [comment]
ah right, but even with pure image batch, like two images it would not work?

File: conversion/deepseek.py [hunk 1]
```diff
@@ -578,6 +578,9 @@ def index_tensors(self, remote_hf_model_id: str | None = None) -> dict[str, Call
     @classmethod
     def filter_tensors(cls, item: tuple[str, Callable[[], Tensor]]) -> tuple[str, Callable[[], Tensor]] | None:
         name, gen = item
+        if (name.startswith(("vision.", "aligner.", "image_"))
```
Comment (CISC) [inline conversion/deepseek.py]:
```suggestion
        if (name.startswith(("aligner.", "image_"))
```
Already here, consider adding the other two:
https://github.com/ggml-org/llama.cpp/blob/9c9a1546d0159781f96e5c1c9ba479e74ecb54c4/conversion/base.py#L1228-L1236

**Comment by ngxson** [comment]
if 2 consecutive images (no text in-between), then they should look exactly like one single image? because input images are already padded, they are aligned to compress ratio, and input batch will contain only embeddings

**Comment by am17an** [comment]
Ah I think M-rope wouldn't work, but this model does not seem to use that. So maybe it's ok

File: conversion/deepseek.py [hunk 2]
```diff
@@ -1018,3 +1021,80 @@ def set_gguf_parameters(self):
...
+            # nn.Linear over flattened (3, p, p) patches == conv2d weight
+            p = self.hparams_vision["patch_size"]
+            data_torch = data_torch.reshape(data_torch.shape[0], 3, p, p)
+
+        if ".attn.wqkv." in name:
+            q, k, v = data_torch.chunk(3, dim=0)
+            yield from super().modify_tensors(q, name.replace("wqkv", "wq"), bid)
+            yield from super().modify_tensors(k, name.replace("wqkv", "wk"), bid)
+            yield from super().modify_tensors(v, name.replace("wqkv", "wv"), bid)
+            return
```
Comment (CISC) [inline conversion/deepseek.py]: Isn't fused QKV supported?

**Review by CISC — APPROVED** [review]

**Comment by ggerganov** [comment]
> Ah I think M-rope wouldn't work, but this model does not seem to use that. So maybe it's ok

So we don't need to resolve this for now?

https://github.com/ggml-org/llama.cpp/blob/1b89a43e3835f0c8bbef5543977151972874a9ce/src/llama-kv-cache-dsv4.cpp#L1610

**Comment by am17an** [comment]
@ggerganov since they're not using M-RoPE I would consider this resolved for now, unless you see some other issue with this (apart from not being able to run unified kv-cache)

**Comment by ggerganov** [comment]
It should probably be OK. Maybe we can add an assert in `llama_kv_cache_dsv4::init_batch` to guarantee that we never receive tokens with the same sequence position - this is what the logic currently assumes.

File: conversion/deepseek.py (hunk 2 above)
Comment (ngxson) [reply conversion/deepseek.py]: fixed in https://github.com/ggml-org/llama.cpp/pull/28133/commits/07335164a1b71e9a4b88688c10791a8698fb4aea

**Comment by ggerganov** [comment]
Do we need some extra changes to support the DSpark? I tried converting and it failed like this: https://github.com/ggml-org/convert/actions/runs/33665921534/job/100367523264#step:5:4003

**Comment by anikifoss** [comment]
> Do we need some extra changes to support the DSpark? I tried converting and it failed like this: https://github.com/ggml-org/convert/actions/runs/33665921534/job/100367523264#step:5:4003

I hit the same error trying to extract --dspark

**Comment by ggerganov** [comment]
@anikifoss See #28294
